design/design.md

```diff
@@ -0,0 +1,56 @@
+# Epiverse search
+
+## User personas
+- Epidemiology tool implementer
+- Epidemiology tool creator
+- Deployment user
+
+## User stories
+This work is mostly targeted at the following persona, with the following stories:
+
+- As an epidemiology tool implementer or epidemiology tool creator I want to find out *if* there is an R package for a specific epidemiological analysis task I have in mind
+- As an epidemiology tool implementer I want to be able to select the right tool for the job.
+- As an epidemiology tool creator I want to find out the implementation details of existing R package for specific epidemiological analysis task I have in mind to determine how to contribute (create a new one, build on existing on).
+- As a deployment user I want to incrementally add/update/delete documents from the search
+- As a deployment user I want to adjust control parameters to influence the quality of the search
+- As a deployment user I want to view the current set of search queries and how they perform for quality control purposes
+
+These user stories have been collected at:
+
+- the '100 days and 100 lines of code' workshop (December 2022)
+- the 'Solution for Collaboratory' launch workshop (January 2023)
+
+## Requirements
+
+The search functionality should:
+
+- take natural language text description of the task the user is trying to perform (e.g., "I want to estimate the risk of a dengue outbreak in Colombia in the next 3 months")
+- return a (list of) valid R package(s) performing the described task, taken from the CRAN Task View in Epidemiology
+- order result by match to the search query
+- provide control parameters to the user to filter the results (e.g., by package name, by package description, by package author)
+
+The search functionality should NOT:
+
+- return hallucinated non-existing R packages
+- return stochastic results. The same query on the same version of the engine should return the same results however the order of results may change. (Perhaps this should be defined probabilistically to allow for future improvements e.g. X% of the time the top result will be the same as the previous run, Y% of the time the top 3 results will be the same as the previous run, etc.)
+- return irrelevant results. There needs to be a threshold of relevance for a package to be returned (again, can be defined probabilistically).
```

Comment:
I don't think the order of results should change. Can you detail the reasoning behind this change?

Reply:
I don't think so too but given the stochastic nature of how LLMs work I would not be surprised for this to happen. I'm trying to make provision should this happen. I'd be more concerned if the relative ordering was dramatically different. Therefore, it makes more sense to accommodate this going in rather than expecting perfectly similar results every time.

Reply:
> I don't think so too but given the stochastic nature of how LLMs work I would not be surprised for this to happen.

It seems that a technical constraint is creeping into the user requirements, which we should avoid:

- Ideally, we should make sure that results are indeed deterministic, even though I understand this is made challenging by the use LLM
- In all cases, even if this was not achieved, we should keep the requirements the same as this indeed what is needed. A note in the final product can be added somewhere if necessary but we shouldn't change our target.